### ISSUE TITLE
XLA duplicates input arguments memory because of wrapped_transpose

### ISSUE THREAD

**Issue by qGentry** [openxla/xla#26987]
Hi, I have simple LLM training step that computes forward+backward pass and updates weights. It is LoRA training, so only very small subset of parameters is trainable (70B total parameters, ~600M trainable). I've checked memory analysis and found that actual memory consumption is way bigger than expected and figured out that this is because XLA is doing some very weird transposition of input weights, resulting in effectively doubling of memory usage for 70B model parameters.

I've tested it on both JAX 0.4.34 and latest version (JAX 0.6.1)

Here's attached HLO dump:

[hlo.tgz](https://github.com/user-attachments/files/20442280/hlo.tgz)

Some of the transposed buffer names:
```
wrapped_transpose.5
input_transpose_fusion.3
wrapped_transpose.10
```

<img width="1628" alt="Image" src="https://github.com/user-attachments/assets/673fd0a8-7645-4ea7-992a-7032d42c445d" />

**Comment by mooskagh** [comment]
I cannot come up with a simple way to fix it (rather than pre-transpose weights), but here's what happens:

Some ops, like dynamic slices, require tensor to be transposed a certain way, for example this one:

```
  ROOT %dynamic-slice.318.1 = bf16[1024,1,29568]{2,1,0} dynamic-slice(%bitcast.780.1, %constant_1683_24, %param_0.1196, %constant_1683_24), dynamic_slice_sizes={1024,1,29568}, metadata={op_name="jit(<unnamed wrapped function>)/jit(main)/while/body/jvp(CausalLMTransformerModel.compute_logits)/transformer/while/body/dynamic_slice" source_file="/usr/local/lib/python3.10/dist-packages/flax/core/axes_scan.py" source_line=179}
```

It's located inside a loop (actually, two nested loops, right?), so optimizer decides to hoist that transposition out of the loop. As a result, the transposed copy of weights have to be alive through the entire duration of the loop. And that's for all weights.

One thing that sometimes helps is AUTO layout (e.g. https://github.com/jax-ml/jax/blob/main/tests/layout_test.py#L36) which would do transposes on the host, but I'm not sure how easy it is to make it work.

**Comment by qGentry** [comment]
Yes, this is indeed a double while loop (gradient accumulation and scan over layers).
We have quite a complex sharding structure so I don't feel like auto layout might work for me.

**Comment by qGentry** [comment]
Here's also an HLO with --xla_dump_hlo_pass_re=.*

[hlo_with_re.tgz](https://github.com/user-attachments/files/20442920/hlo_with_re.tgz)

**Comment by qGentry** [comment]
Yeah, the issue with previous HLO was in compilation cache. Here's one with it. Couldn't attach it to github because of file size restrictions.
https://drive.google.com/file/d/1H_BCJ7vojqWQM2wbOVlLOB_k-NQUxj20/view?usp=sharing

**Comment by jreiffers** [comment]
This appears to be caused by layout assignment, which introduces a transposing copy in the top level. Not sure yet why that's happening, we're having some trouble with VLOGing.

**Comment by jreiffers** [comment]
AFAICT, the reason are all these all-gathers with dimension=1, like this:

```
%all-gather.41 = bf16[1,64,8192]{2,1,0} all-gather(%convert.442), channel_id=61, replica_groups=[1,8]<=[8], dimensions={1}, use_global_device_ids=true
```

Because the all-gather dimension has to be major, gpu_layout_assignment thinks that they need to get a layout of `{2,0,1}`, which then gets propagated everywhere. But I don't think that's actually true, because the shape is `[1,64,8192]`, i.e. dim 0 is degenerate, so `{2,1,0}` should work as well and everything should be fine.

@mooskagh I see a couple of ways to fix this, e.g. by getting rid of these degenerate dims or by attempting to make gpu_layout_assignment smarter. The former seems easier to me, WDYT?

**Comment by jreiffers** [comment]
https://github.com/openxla/xla/commit/4c817c55329fc4139451db14aea2d66bab737496 This gets rid of most of the copies, a handful are still there. Looking into the remaining ones now.

**Comment by jreiffers** [comment]
The remaining copies are actually needed, because the all-gathers are gathering along a minor dimension:

![Image](https://github.com/user-attachments/assets/7e12f87e-e3c5-4407-927a-c9e35b1ffd51)
